### PR TITLE
Swap contract with ver1 address conversion support

### PR DESCRIPTION
**Pull request by psy2848048**
Not finalized. Just for interim review for swap contract

### Point
* Unnecessary objects
* Unnecessary method
* Any improvement especially for `named_key` manipulations

### CODE REVIEW

File: execution-engine/types/src/system_contract_errors/mint.rs [hunk 1]
```diff
@@ -38,6 +38,21 @@ pub enum Error {
...
+    ExceededSwapAllowanceByKyc = 9,
+    /// Insufficient number of swap parameters
+    #[fail(display = "Insufficient number of swap parameters")]
+    InsufficientNumOfSwapParams = 10,
+    /// Caller has not registered KYC
+    #[fail(display = "Caller has not registered KYC")]
+    NotRegisteredKYC = 11,
+    /// Invalid KYC level value
+    #[fail(display = "Invalid KYC level value")]
+    InvalidKYCLevelValue = 12,
```
Comment (jhyeom26) [inline execution-engine/types/src/system_contract_errors/mint.rs]: Would be better to define them as `ApiError::User` in swap-logic contract. `mint::Error` is preserved for `hdac_mint_token` contract.

File: execution-engine/contracts/swap/swap-logic/src/swap_control/mod.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
+mod signiture_verification;
+mod storage;
```
Comment (jhyeom26) [inline execution-engine/contracts/swap/swap-logic/src/swap_control/mod.rs]: This mod name might be confusing.

File: execution-engine/contracts/swap/swap-logic/src/swap_control/mod.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+};
+
+use crate::constants;
+use signiture_verification::signature_verification;
+
+// Admin features
+
+pub fn set_swap_hash(swap_hash: Key) {
+    if runtime::get_caller() != storage::load_admin() {
+        runtime::revert(ApiError::NoAccessRights);
```
Comment (jhyeom26) [inline execution-engine/contracts/swap/swap-logic/src/swap_control/mod.rs]: `ApiError::NoAccessRight` represents attempting to use of unauthorized uref. Would be more exact that defining swap's own error system.

File: execution-engine/contracts/swap/swap-proxy/src/client_api/mod.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    UpdateKYCLevel(PublicKey, U512),
+    UpdateStatusSwapableTokenSent(PublicKey, U512),
+    UpdateKYCStep(PublicKey, U512),
+    GetToken(Key, Vec<String>, Vec<String>, Vec<String>, Vec<String>),
+}
+
+fn get_contract_ref() -> ContractRef {
+    let contract_hash = runtime::get_key(method_names::proxy::NAME_SWAP_HASH)
+        .unwrap_or_revert_with(ApiError::GetKey);
+    contract_hash.to_contract_ref().unwrap_or_revert()
```
Comment (jhyeom26) [inline execution-engine/contracts/swap/swap-proxy/src/client_api/mod.rs:52]: These can be simplified like below,
```Rust
fn get_contract_ref() -> ContractRef {
    runtime::get_key(method_names::proxy::NAME_SWAP_HASH)
        .unwrap_or_revert_with(ApiError::GetKey);
        .to_contract_ref()
        .unwrap_or_revert()
}
```

Also, this proxy requires that the caller has the NAME_SWAP_HASH in their named key.
This requires a user to run the contract storing NAME_SWAP_HASH in their named_key.
Why don't we allow a user just to call by hash without storing it under their named_key?

More than that, how about integrating the logic, proxy, and install into a single contract?

File: execution-engine/types/src/system_contract_errors/mint.rs (hunk 1 above)
Comment (psy2848048) [reply execution-engine/types/src/system_contract_errors/mint.rs]: https://github.com/hdac-io/CasperLabs/pull/29/commits/1fa91df46f1a0e642ab4d6c746ba7a9c5c1fd5fc

File: execution-engine/contracts/swap/swap-logic/src/swap_control/mod.rs (hunk 2 above)
Comment (psy2848048) [reply execution-engine/contracts/swap/swap-logic/src/swap_control/mod.rs]: https://github.com/hdac-io/CasperLabs/pull/29/commits/1fa91df46f1a0e642ab4d6c746ba7a9c5c1fd5fc

File: execution-engine/contracts/swap/swap-logic/src/swap_control/mod.rs (hunk 3 above)
Comment (psy2848048) [reply execution-engine/contracts/swap/swap-logic/src/swap_control/mod.rs]: https://github.com/hdac-io/CasperLabs/pull/29/commits/1fa91df46f1a0e642ab4d6c746ba7a9c5c1fd5fc

File: execution-engine/contracts/swap/swap-proxy/src/client_api/mod.rs (hunk 4 above)
Comment (psy2848048) [reply execution-engine/contracts/swap/swap-proxy/src/client_api/mod.rs:52]: > Why don't we allow a user just to call by hash without storing it under their named_key?

Do you mean to get hash as a parameter in every contract call?

Comment (psy2848048) [reply execution-engine/contracts/swap/swap-proxy/src/client_api/mod.rs:52]: > More than that, how about integrating the logic, proxy, and install into a single contract?

Try to do that but assign to lower priority.

**Pull request closed without merging**